sdks/go/pkg/beam/artifact/materialize.go

```diff
@@ -69,41 +74,100 @@ func newMaterialize(ctx context.Context, endpoint string, dependencies []*pipepb
 }
 
-func newMaterializeWithClient(ctx context.Context, client jobpb.ArtifactRetrievalServiceClient, dependencies []*pipepb.ArtifactInformation, dest string) ([]*jobpb.ArtifactMetadata, error) {
+func newMaterializeWithClient(ctx context.Context, client jobpb.ArtifactRetrievalServiceClient, dependencies []*pipepb.ArtifactInformation, dest string) ([]*pipepb.ArtifactInformation, error) {
 	resolution, err := client.ResolveArtifacts(ctx, &jobpb.ResolveArtifactsRequest{Artifacts: dependencies})
 	if err != nil {
 		return nil, err
 	}
 
-	var md []*jobpb.ArtifactMetadata
+	var artifacts []*pipepb.ArtifactInformation
 	var list []retrievable
 	for _, dep := range resolution.Replacements {
 		path, err := extractStagingToPath(dep)
 		if err != nil {
 			return nil, err
 		}
-		md = append(md, &jobpb.ArtifactMetadata{
-			Name: path,
+		var filePayload pipepb.ArtifactFilePayload
+		if dep.TypeUrn != URNFileArtifact {
+			filePayload = pipepb.ArtifactFilePayload{
+				Path: path,
+			}
+		} else {
+			typePayload := pipepb.ArtifactFilePayload{}
+			if err := proto.Unmarshal(dep.TypePayload, &typePayload); err != nil {
+				return nil, errors.Wrap(err, "failed to parse artifact file payload")
+			}
+			filePayload = pipepb.ArtifactFilePayload{
+				Path:   path,
+				Sha256: typePayload.Sha256,
+			}
+		}
+		newTypePayload, err := proto.Marshal(&filePayload)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create artifact type payload")
+		}
+		artifacts = append(artifacts, &pipepb.ArtifactInformation{
+			TypeUrn:     URNFileArtifact,
+			TypePayload: newTypePayload,
+			RoleUrn:     dep.RoleUrn,
+			RolePayload: dep.RolePayload,
 		})
 
+		rolePayload, err := proto.Marshal(&pipepb.ArtifactStagingToRolePayload{
+			StagedName: path,
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create artifact role payload")
+		}
 		list = append(list, &artifact{
 			client: client,
-			dep:    dep,
+			dep: &pipepb.ArtifactInformation{
+				TypeUrn:     dep.TypeUrn,
+				TypePayload: dep.TypePayload,
+				RoleUrn:     URNStagingTo,
+				RolePayload: rolePayload,
+			},
 		})
 	}
 
-	return md, MultiRetrieve(ctx, 10, list, dest)
+	return artifacts, MultiRetrieve(ctx, 10, list, dest)
+}
+
+// Used for generating unique IDs. We assign uniquely generated names to staged files without staging names.
+var idCounter uint64
+
+func generateId() string {
+	id := atomic.AddUint64(&idCounter, 1)
+	return strconv.FormatUint(id, 10)
 }
 
 func extractStagingToPath(artifact *pipepb.ArtifactInformation) (string, error) {
-	if artifact.RoleUrn != URNStagingTo {
-		return "", errors.Errorf("Unsupported artifact role %s", artifact.RoleUrn)
+	var stagedName string
+	if artifact.RoleUrn == URNStagingTo {
+		role := pipepb.ArtifactStagingToRolePayload{}
+		if err := proto.Unmarshal(artifact.RolePayload, &role); err != nil {
+			return "", err
+		}
+		stagedName = role.StagedName
+	} else {
```

Comment:
Guard this with artifact.TypeUrn is URNFileArtifact? And a final else that's an error?

Reply:
Done. That looks better. Thanks.